### PR TITLE
New authority for Discogs data.

### PR DESCRIPTION
**Pull request by tworrall**

### CODE REVIEW

File: config/discogs-formats.yml [hunk 1]
```diff
@@ -0,0 +1,346 @@
+# carriers
```
Comment (elrayle) [inline config/discogs-formats.yml]: This file should be in lib/generators/qa/discogs/templates/config only.  There should be instructions on how to run the generator that will put this in an app.

File: config/discogs-genres.yml [hunk 2]
```diff
@@ -0,0 +1,627 @@
+Acid:
+  label: Psychedelic rock music
```
Comment (elrayle) [inline config/discogs-genres.yml]: This file should be in lib/generators/qa/discogs/templates/config only.  There should be instructions on how to run the generator that will put this in an app.

File: lib/qa/authorities/discogs.rb [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+    def initialize
+      self.primary_artists = []
+    end
+
+    def search(q, tc)
+      # Discogs distinguishes between masters and releases, where a release represents a specific
+      # physical or digital object and a master represents a set of similar releases. Use of a
+      # subauthority (e.g., /qa/search/discogs/master) will target a specific type. No subauthority
+      # will search for both types.
```
Comment (elrayle) [inline lib/qa/authorities/discogs.rb]: Appreciate the comments.  Makes things easier to follow.

File: lib/qa/authorities/discogs.rb [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    include DiscogsTranslation
+
+    class_attribute :discogs_secret, :discogs_key
+    attr_accessor :primary_artists
+
+    def initialize
+      self.primary_artists = []
+    end
+
+    def search(q, tc)
```
Comment (elrayle) [inline lib/qa/authorities/discogs.rb]: Can you add yardoc for parameters?  I know existing code is not fully documented, but the standard going forward is to include yardoc for new code.

https://www.rubydoc.info/gems/yard/file/docs/Tags.md#List_of_Available_Tags

This file is well documented and can serve as an example.

File: config/discogs-formats.yml (hunk 1 above)
Comment (tworrall) [reply config/discogs-formats.yml]: I was following the examples of other authorities. These files are all in both locations, though it wasn't clear why:

- authorities.yml
- states.yml
- oclcts-authorities.yml

File: lib/qa/authorities/discogs.rb (hunk 4 above)
Comment (elrayle) [reply lib/qa/authorities/discogs.rb]: Did you look at using AuthorityWithSubAuthority?

File: lib/qa/authorities/discogs.rb [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+      # If the requested format is json-ld, call the build_graph method in the translation module;
+      # otherwise, just return the Discogs json. If there's no subauthority, check the response
+      # to determine if it should go to the translation module.
+      response = tc.params["subauthority"].present? ? json(find_url(id, tc.params["subauthority"])) : fetch_discogs_results(id)
+      return build_graph(response) unless tc.params["format"] != "jsonld" || response["message"].present?
+      response
+    end
+
+    def build_query_url(q, subauthority)
+      escaped_q = URI.escape(q)
```
Comment (elrayle) [inline lib/qa/authorities/discogs.rb]: The use of `URI.escape` is deprecated.  A recent [PR](https://github.com/samvera/questioning_authority/pull/210) switched to using `ERB::Util.url_encode`.

File: lib/qa/authorities/discogs/discogs_instance_builder.rb [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        descs.each do |desc|
+          # map discogs description field to the corresponding LOC type
+          df = discogs_formats[desc.gsub(/\s+/, "")]
+          if df.present?
+            stmts += build_format_characteristics(df, count)
+          else
+            stmts << contruct_stmt_literal_object("Instance1", "http://id.loc.gov/ontologies/bibframe/editionStatement", desc)
+          end
+        end
+        stmts # w/out this line, building the graph throws an undefined method `graph_name=' error
```
Comment (elrayle) [inline lib/qa/authorities/discogs/discogs_instance_builder.rb]: If descs is an array, descs.each will return the descs array.  So it makes sense to me that you would need to return stmts.

File: lib/qa/authorities/discogs/discogs_instance_builder.rb [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+      def build_base_materials(name)
+        stmts = []
+        return stmts unless name == "Vinyl" || name == "Shellac"
+        id = name == "Vinyl" ? "300014502" : "300014918"
+
+        stmts << contruct_stmt_uri_object("Instance1", "http://id.loc.gov/ontologies/bibframe/baseMaterial", "http://vocab.getty.edu/aat/" + id)
+        stmts << contruct_stmt_uri_object("http://vocab.getty.edu/aat/" + id, rdf_type_predicate, "http://id.loc.gov/ontologies/bibframe/BaseMaterial")
+        stmts << contruct_stmt_literal_object("http://vocab.getty.edu/aat/" + id, rdfs_label_predicate, name)
+        stmts # w/out this line, building the graph throws an undefined method `graph_name=' error
```
Comment (elrayle) [inline lib/qa/authorities/discogs/discogs_instance_builder.rb]: This one I'm surprised by.  The `<<` operator on an array returns the updated array.  But its fine to have stmts here.  It makes it clear that stmts is being returned.

File: config/discogs-formats.yml (hunk 1 above)
Comment (tworrall) [reply config/discogs-formats.yml]: done

File: config/discogs-genres.yml (hunk 2 above)
Comment (tworrall) [reply config/discogs-genres.yml]: done

File: lib/qa/authorities/discogs.rb (hunk 4 above)
Comment (tworrall) [reply lib/qa/authorities/discogs.rb]: done

File: lib/qa/authorities/discogs.rb (hunk 5 above)
Comment (tworrall) [reply lib/qa/authorities/discogs.rb]: done

File: lib/qa/authorities/discogs/discogs_instance_builder.rb (hunk 6 above)
Comment (tworrall) [reply lib/qa/authorities/discogs/discogs_instance_builder.rb]: removed the closing line

File: lib/qa/authorities/discogs/discogs_instance_builder.rb (hunk 7 above)
Comment (tworrall) [reply lib/qa/authorities/discogs/discogs_instance_builder.rb]: Removed the closing line.

File: lib/qa/authorities/discogs/generic_authority.rb [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+      #     ],
+      #     "Formats": [
+      #       "Shellac",
+      #       "10\"",
+      #       "78 RPM"
+      #     ],
+      #     "Type": [
+      #       "release"
+      #     ]
+      #   }
```
Comment (elrayle) [inline lib/qa/authorities/discogs/generic_authority.rb]: nice including an example!

**Pull request merged**